In [ ]:
!pip install tensorflow

In [ ]:
!pip install --upgrade keras

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 60.7 MB/s eta 0:00:00
  Attempting uninstall: keras
    Found existing installation: keras 3.8.0
    Uninstalling keras-3.8.0:
      Successfully uninstalled keras-3.8.0


In [ ]:
!pip install datasets

In [ ]:
!huggingface-cli login

⚠️  Warning: 'huggingface-cli login' is deprecated. Use 'hf auth login' instead.

    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

    To log in, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Enter your token (input will not be visible): 
Add token as git credential? (Y/n) n
Token is valid (permission: write).
The token `BigEarthNetFinal` has been saved to /root/.cache/huggingface/stored_tokens
Your token has been saved to /root

In [ ]:
#Download BigEarthNet dataset
from datasets import load_dataset
ds = load_dataset("danielz01/BigEarthNet-S2-v1.0")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/1.04k [00:00<?, ?B/s]

s2-rgb/test-00000-of-00007-fa75b96ba0241(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/test-00001-of-00007-7746faa4c2987(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/test-00002-of-00007-3df6bac341e50(…):   0%|          | 0.00/496M [00:00<?, ?B/s]

s2-rgb/test-00003-of-00007-134378bdb2c97(…):   0%|          | 0.00/472M [00:00<?, ?B/s]

s2-rgb/test-00004-of-00007-ba56718a154a3(…):   0%|          | 0.00/477M [00:00<?, ?B/s]

s2-rgb/test-00005-of-00007-db24561286ec3(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/test-00006-of-00007-2d2952dd94da4(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/val-00000-of-00007-f0643798602cca(…):   0%|          | 0.00/455M [00:00<?, ?B/s]

s2-rgb/val-00001-of-00007-04b0d699da1eb9(…):   0%|          | 0.00/455M [00:00<?, ?B/s]

s2-rgb/val-00002-of-00007-fdf80718140deb(…):   0%|          | 0.00/486M [00:00<?, ?B/s]

s2-rgb/val-00003-of-00007-e34a5e008b9d45(…):   0%|          | 0.00/465M [00:00<?, ?B/s]

s2-rgb/val-00004-of-00007-d7d1605cae3e62(…):   0%|          | 0.00/470M [00:00<?, ?B/s]

s2-rgb/val-00005-of-00007-67da670d67109d(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/val-00006-of-00007-e8aa45ccad7d0d(…):   0%|          | 0.00/480M [00:00<?, ?B/s]

s2-rgb/train-00000-of-00015-d6560d607332(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/train-00001-of-00015-9a124129340e(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00002-of-00015-288e311ae08e(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00003-of-00015-1ccc32c48471(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00004-of-00015-36361aea05da(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/train-00005-of-00015-90829c95f348(…):   0%|          | 0.00/501M [00:00<?, ?B/s]

s2-rgb/train-00006-of-00015-3c4a1537f29b(…):   0%|          | 0.00/480M [00:00<?, ?B/s]

s2-rgb/train-00007-of-00015-11e2fb6c06dd(…):   0%|          | 0.00/467M [00:00<?, ?B/s]

s2-rgb/train-00008-of-00015-34d29c632d99(…):   0%|          | 0.00/465M [00:00<?, ?B/s]

s2-rgb/train-00009-of-00015-d2d8d6d6d458(…):   0%|          | 0.00/473M [00:00<?, ?B/s]

s2-rgb/train-00010-of-00015-ca2886dc8cd5(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/train-00011-of-00015-9a1860b8270f(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/train-00012-of-00015-2eacca1c13d3(…):   0%|          | 0.00/492M [00:00<?, ?B/s]

s2-rgb/train-00013-of-00015-d522c508701a(…):   0%|          | 0.00/481M [00:00<?, ?B/s]

s2-rgb/train-00014-of-00015-df0029906d78(…):   0%|          | 0.00/494M [00:00<?, ?B/s]

Generating test split:   0%|          | 0/125866 [00:00<?, ? examples/s]

Generating val split:   0%|          | 0/123723 [00:00<?, ? examples/s]

Generating train split:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
#This method converts the image labels from the 43-class version to the 19-class version
def map_labels(sample):
    label = sample["labels"]
    for i in range(len(label)):
      if label[i] == "Continuous urban fabric":
        label[i] = "Urban fabric"
      if label[i] =="Discontinuous urban fabric":
        label[i] = "Urban fabric"
      if label[i] == "Non-irrigated arable land":
        label[i] = "Arable land"
      if label[i] == "Permanently irrigated land":
        label[i] = "Arable land"
      if label[i] == "Rice fields":
        label[i] = "Arable land"
      if label[i] == "Vineyards":
        label[i] = "Permanent crops"
      if label[i] == "Fruit trees and berry plantations":
        label[i] = "Permanent crops"
      if label[i] == "Olive groves":
        label[i] = "Permanent crops"
      if label[i] == "Annual crops associated with permanent crops":
        label[i] = "Permanent crops"
      if label[i] == "Natural grassland":
        label[i] = "Natural grassland and sparsely vegetated areas"
      if label[i] == "Sparsely vegetated areas":
        label[i] = "Natural grassland and sparsely vegetated areas"
      if label[i] == "Moors and heathland":
        label[i] = "Moors, heathland and sclerophyllous vegetation"
      if label[i] == "Sclerophyllous vegetation":
        label[i] = "Moors, heathland and sclerophyllous vegetation"
      if label[i] == "Inland marshes":
        label[i] = "Inland wetlands"
      if label[i] == "Peatbogs":
        label[i] = "Inland wetlands"
      if label[i] == "Salt marshes":
        label[i] = "Coastal wetlands"
      if label[i] == "Salines":
        label[i] = "Coastal wetlands"
      if label[i] == "Water courses":
        label[i] = "Inland waters"
      if label[i] == "Water bodies":
        label[i] = "Inland waters"
      if label[i] == "Coastal lagoons":
        label[i] = "Marine waters"
      if label[i] == "Estuaries":
        label[i] = "Marine waters"
      if label[i] == "Sea and ocean":
        label[i] = "Marine waters"
      if label[i] == "Transitional woodland/shrub":
        label[i] = "Transitional woodland, shrub"

    if "Road and rail networks and associated land" in label:
      label.remove("Road and rail networks and associated land")
    if "Sport and leisure facilities" in label:
      label.remove("Sport and leisure facilities")
    if "Port areas" in label:
      label.remove("Port areas")
    if "Airports" in label:
      label.remove("Airports")
    if "Mineral extraction sites" in label:
      label.remove("Mineral extraction sites")
    if "Dump sites" in label:
      label.remove("Dump sites")
    if "Construction sites" in label:
      label.remove("Construction sites")
    if"Green urban areas" in label:
      label.remove("Green urban areas")
    if"Bare rock" in label:
      label.remove("Bare rock")
    if"Burnt areas" in label:
      label.remove("Burnt areas")
    if"Intertidal flats" in label:
      label.remove("Intertidal flats")

    if len(set(label)) < len(label):
      label = set(label)
    if len(label) == 0:
      print(label)

    return sample

In [ ]:
ds["train"] = ds["train"].map(map_labels) # Convert Traning Set to 19-class version

Map:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
ds["val"] = ds["val"].map(map_labels) # Convert Validation Set to 19-class version

Map:   0%|          | 0/123723 [00:00<?, ? examples/s]

In [ ]:
ds['test'] = ds['test'].map(map_labels) # Convert Test set to 19-class version

Map:   0%|          | 0/125866 [00:00<?, ? examples/s]

In [ ]:
# This method hot-one encodes labels
encoder_dict = {
        "Urban fabric": 0,
        "Industrial or commercial units": 1,
        "Arable land": 2,
        "Permanent crops": 3,
        "Pastures": 4,
        "Complex cultivation patterns": 5,
        "Land principally occupied by agriculture, with significant areas of natural vegetation": 6,
        "Agro-forestry areas": 7,
        "Broad-leaved forest": 8,
        "Coniferous forest": 9,
        "Mixed forest": 10,
        "Natural grassland and sparsely vegetated areas": 11,
        "Moors, heathland and sclerophyllous vegetation": 12,
        "Transitional woodland, shrub": 13,
        "Beaches, dunes, sands": 14,
        "Inland wetlands": 15,
        "Coastal wetlands": 16,
        "Inland waters": 17,
        "Marine waters": 18
}


def hotOne(sample):
  new_label = [0] * 19
  label = sample["labels"]
  for i in label:
    new_label[encoder_dict[i]] = 1
  sample["labels"] = new_label
  return sample

In [ ]:
ds["train"] = ds["train"].map(hotOne) # Hot-one encode training set

Map:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
ds["val"] = ds["val"].map(hotOne) # Hot-one encode validation set

Map:   0%|          | 0/123723 [00:00<?, ? examples/s]

In [ ]:
ds["test"] = ds["test"].map(hotOne) # Hot-one encode test set

Map:   0%|          | 0/125866 [00:00<?, ? examples/s]

In [ ]:
print (ds['train'][0])

{'img': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=120x120 at 0x7948FA8BD010>, 'labels': [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'coordinates': {'lrx': 634800, 'lry': 5994420, 'ulx': 633600, 'uly': 5995620}, 'projection': 'PROJCS["WGS 84 / UTM zone 29N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",-9],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32629"]]', 'tile_source': 'S2A_MSIL1C_20170717T113321_N0205_R080_T29UPA_20170717T113326.SAFE', 'acquisition_date'

In [ ]:
from keras import layers
import keras

import matplotlib.pyplot as plt
import numpy as np

from tensorflow.keras import layers
import tensorflow.keras.backend as ops
import tensorflow

import matplotlib.pyplot as plt
import numpy as np


#CCT Hyerparameters

#Positional Embedding
positional_emb = True
#Two convultional Layers
conv_layers = 2
#Projection dimension
projection_dim = 256
#Number of Transformer heads
num_heads = 4

transformer_units = [
    projection_dim,
    projection_dim,
]
#Number of Transformer layers
transformer_layers = 6
#Stochastic Depth Rate
stochastic_depth_rate = 0.1
#Learning Rate
learning_rate = 0.001
#Weight Decay
weight_decay = 0.0001
#Batch size
batch_size = 128 # Reduced batch size to mitigate memory issues
#number
num_epochs = 30
#For Random Crop
image_size = 120
#Number of Classes
num_classes = 19
#Image shape
input_shape = (120, 120, 3)

In [ ]:
#All CCT methods were imported from Keras
#https://keras.io/examples/vision/cct/

In [ ]:
class CCTTokenizer(layers.Layer):
    def __init__(
        self,
        kernel_size=3,
        stride=1,
        padding=1,
        pooling_kernel_size=3,
        pooling_stride=2,
        num_conv_layers=conv_layers,
        num_output_channels=[64, 256], # Changed from [64, 128] to [64, 256]
        positional_emb=positional_emb,
        **kwargs,
    ):
        super().__init__(**kwargs)

        # This is our tokenizer.
        self.conv_model = keras.Sequential()
        for i in range(num_conv_layers):
            self.conv_model.add(
                layers.Conv2D(
                    num_output_channels[i],
                    kernel_size,
                    stride,
                    padding="valid",
                    use_bias=False,
                    activation="relu",
                    kernel_initializer="he_normal",
                )
            )
            self.conv_model.add(layers.ZeroPadding2D(padding))
            self.conv_model.add(
                layers.MaxPooling2D(pooling_kernel_size, pooling_stride, "same")
            )

        self.positional_emb = positional_emb

    def call(self, images):
        outputs = self.conv_model(images)
        # After passing the images through our mini-network the spatial dimensions
        # are flattened to form sequences.
        reshaped = keras.ops.reshape(
            outputs,
            (
                -1,
                keras.ops.shape(outputs)[1] * keras.ops.shape(outputs)[2],
                keras.ops.shape(outputs)[-1],
            ),
        )
        return reshaped

In [ ]:
class PositionEmbedding(keras.layers.Layer):
    def __init__(
        self,
        sequence_length,
        initializer="glorot_uniform",
        **kwargs,
    ):
        super().__init__(**kwargs)
        if sequence_length is None:
            raise ValueError("`sequence_length` must be an Integer, received `None`.")
        self.sequence_length = int(sequence_length)
        self.initializer = keras.initializers.get(initializer)

    def get_config(self):
        config = super().get_config()
        config.update(
            {
                "sequence_length": self.sequence_length,
                "initializer": keras.initializers.serialize(self.initializer),
            }
        )
        return config

    def build(self, input_shape):
        feature_size = input_shape[-1]
        self.position_embeddings = self.add_weight(
            name="embeddings",
            shape=[self.sequence_length, feature_size],
            initializer=self.initializer,
            trainable=True,
        )

        super().build(input_shape)

    def call(self, inputs, start_index=0):
        shape = keras.ops.shape(inputs)
        feature_length = shape[-1]
        sequence_length = shape[-2]
        # trim to match the length of the input sequence, which might be less
        # than the sequence_length of the layer.
        position_embeddings = keras.ops.convert_to_tensor(self.position_embeddings)
        position_embeddings = keras.ops.slice(
            position_embeddings,
            (start_index, 0),
            (sequence_length, feature_length),
        )
        return keras.ops.broadcast_to(position_embeddings, shape)

    def compute_output_shape(self, input_shape):
        return input_shape

In [ ]:
  class SequencePooling(layers.Layer):
    def __init__(self):
        super().__init__()
        self.attention = layers.Dense(1)

    def call(self, x):
        attention_weights = keras.ops.softmax(self.attention(x), axis=1)
        attention_weights = keras.ops.transpose(attention_weights, axes=(0, 2, 1))
        weighted_representation = keras.ops.matmul(attention_weights, x)
        return keras.ops.squeeze(weighted_representation, -2)

In [ ]:
# Referred from: github.com:rwightman/pytorch-image-models.
class StochasticDepth(layers.Layer):
    def __init__(self, drop_prop, **kwargs):
        super().__init__(**kwargs)
        self.drop_prob = drop_prop
        self.seed_generator = keras.random.SeedGenerator(1337)

    def call(self, x, training=None):
        if training:
            keep_prob = 1 - self.drop_prob
            shape = (keras.ops.shape(x)[0],) + (1,) * (len(x.shape) - 1)
            random_tensor = keep_prob + keras.random.uniform(
                shape, 0, 1, seed=self.seed_generator
            )
            random_tensor = keras.ops.floor(random_tensor)
            return (x / keep_prob) * random_tensor
        return x

In [ ]:
def mlp(x, hidden_units, dropout_rate):
    for units in hidden_units:
        x = layers.Dense(units, activation=keras.ops.gelu)(x)
        x = layers.Dropout(dropout_rate)(x)
    return x

In [ ]:
#Note the rescaling layer. These layers have pre-defined inference behavior.
data_augmentation = keras.Sequential(
    [
        layers.Rescaling(scale=1.0 / 255),
        layers.RandomFlip("horizontal"),
    ],
    name="data_augmentation",
)

In [ ]:
def create_cct_model(
    image_size=image_size,
    input_shape=input_shape,
    num_heads=num_heads,
    projection_dim=projection_dim,
    transformer_units=transformer_units,
):
    inputs = layers.Input(input_shape)

    # Augment data.
    augmented = data_augmentation(inputs)

    # Encode patches.
    cct_tokenizer = CCTTokenizer()
    encoded_patches = cct_tokenizer(augmented)

    # Apply positional embedding.
    if positional_emb:
        sequence_length = encoded_patches.shape[1]
        encoded_patches += PositionEmbedding(sequence_length=sequence_length)(
            encoded_patches
        )

    # Calculate Stochastic Depth probabilities.
    dpr = [x for x in np.linspace(0, stochastic_depth_rate, transformer_layers)]

    # Create multiple layers of the Transformer block.
    for i in range(transformer_layers):
        # Layer normalization 1.
        x1 = layers.LayerNormalization(epsilon=1e-5)(encoded_patches)

        # Create a multi-head attention layer.
        attention_output = layers.MultiHeadAttention(
            num_heads=num_heads, key_dim=projection_dim, dropout=0.1
        )(x1, x1)

        # Skip connection 1.
        attention_output = StochasticDepth(dpr[i])(attention_output)
        x2 = layers.Add()([attention_output, encoded_patches])

        # Layer normalization 2.
        x3 = layers.LayerNormalization(epsilon=1e-5)(x2)

        # MLP.
        x3 = mlp(x3, hidden_units=transformer_units, dropout_rate=0.1)

        # Skip connection 2.
        x3 = StochasticDepth(dpr[i])(x3)
        encoded_patches = layers.Add()([x3, x2])

    # Apply sequence pooling.
    representation = layers.LayerNormalization(epsilon=1e-5)(encoded_patches)
    weighted_representation = SequencePooling()(representation)

    # Classify outputs.
    logits = layers.Dense(num_classes)(weighted_representation)
    # Create the Keras model.
    model = keras.Model(inputs=inputs, outputs=logits)
    return model

In [ ]:
#Create CCT model
model = create_cct_model()

#Create Adam Optimizer with learning Rate and Weight Decay
optimizer = keras.optimizers.AdamW(learning_rate=0.001, weight_decay=0.0001)

#Compile
model.compile(
    optimizer=optimizer,
    loss=keras.losses.BinaryCrossentropy(
        from_logits=True, label_smoothing=0.1
    ),
    metrics=[
        keras.metrics.CategoricalAccuracy(name="accuracy"),
        keras.metrics.TopKCategoricalAccuracy(5, name="top-5-accuracy"),
    ],
)


In [ ]:
model.summary()

Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 120, 120,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ data_augmentation   │ (None, 120, 120,  │          0 │ input_layer[0][0] │
│ (Sequential)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ cct_tokenizer       │ (None, 900, 256)  │    149,184 │ data_augmentatio… │
│ (CCTTokenizer)      │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ position_embedding  │ (None, 900, 256)  │    230,400 │ cct_tokenizer[0]… │
│ (PositionEmbedding) │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add (Add)           │ (None, 900, 256)  │          0 │ cct_tokenizer[0]… │
│                     │                   │            │ position_embeddi… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalization │ (None, 900, 256)  │        512 │ add[0][0]         │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 256)  │  1,051,904 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth    │ (None, 900, 256)  │          0 │ multi_head_atten… │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_1 (Add)         │ (None, 900, 256)  │          0 │ stochastic_depth… │
│                     │                   │            │ add[0][0]         │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 256)  │        512 │ add_1[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 900, 256)  │     65,792 │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 900, 256)  │          0 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 900, 256)  │     65,792 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 900, 256)  │          0 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth_1  │ (None, 900, 256)  │          0 │ dropout_2[0][0]   │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_2 (Add)         │ (None, 900, 256)  │          0 │ stochastic_depth… │
│                     │                   │            │ add_1[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 256)  │        512 │ add_2[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 256)  │  1,051,904 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat

 Total params: 7,492,308 (28.58 MB)

 Trainable params: 7,492,308 (28.58 MB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
checkpoint_filepath = "'/content/drive/MyDrive/BigEarthFinal/checkpoint.weights.h5" #Create Checkpoint

In [ ]:
#Create Checkpoint call back
checkpoint_callback = keras.callbacks.ModelCheckpoint(
    checkpoint_filepath,
    monitor="val_accuracy",
    save_best_only=True,
    save_weights_only=True,
)

In [ ]:
tf_ds = ds["train"].to_tf_dataset( #Create image generator for Training Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=batch_size,
    num_workers = 4, #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU
    #prefetch = True
)

/usr/local/lib/python3.11/dist-packages/datasets/arrow_dataset.py:403: FutureWarning: The output of `to_tf_dataset` will change when a passing single element list for `labels` or `columns` in the next datasets version. To return a tuple structure rather than dict, pass a single string.
Old behaviour: columns=['a'], labels=['labels'] -> (tf.Tensor, tf.Tensor)  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor)  
New behaviour: columns=['a'],labels=['labels'] -> ({'a': tf.Tensor}, {'labels': tf.Tensor})  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor) 
  warnings.warn(


In [ ]:
tf_ds_val = ds["val"].to_tf_dataset( #Create image generator for Validation Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=batch_size,
    num_workers = 4 #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU

)

In [ ]:
#ds["train"][0]["img"]

In [ ]:
# Fit model with 30 epochs and a batch size of 256
history = model.fit(
    tf_ds,
    batch_size=batch_size,
    epochs = 30,
    callbacks=[checkpoint_callback],
    validation_data = tf_ds_val,

)


Epoch 1/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 1107s 492ms/step - accuracy: 0.4573 - loss: 0.3604 - top-5-accuracy: 0.7981 - val_accuracy: 0.4858 - val_loss: 0.3356 - val_top-5-accuracy: 0.8680
Epoch 2/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 979s 462ms/step - accuracy: 0.4845 - loss: 0.3301 - top-5-accuracy: 0.8662 - val_accuracy: 0.4707 - val_loss: 0.3216 - val_top-5-accuracy: 0.8642
Epoch 3/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 979s 462ms/step - accuracy: 0.4874 - loss: 0.3214 - top-5-accuracy: 0.8804 - val_accuracy: 0.4893 - val_loss: 0.3158 - val_top-5-accuracy: 0.8839
Epoch 4/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 980s 463ms/step - accuracy: 0.4896 - loss: 0.3161 - top-5-accuracy: 0.8889 - val_accuracy: 0.4943 - val_loss: 0.3126 - val_top-5-accuracy: 0.8910
Epoch 5/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 980s 463ms/step - accuracy: 0.4912 - loss: 0.3124 - top-5-accuracy: 0.8947 - val_accuracy: 0.4978 - val_loss: 0.3091 - val_top-5-accuracy: 0.8953
Epoch 6/30
2107/2107 ━━━━━━━━━━━━━━━━━━━━ 980s 463ms/step -

In [ ]:
file_path = "/content/drive/MyDrive/epoch1.weights.h5" #Create file path for firest set of weights

In [ ]:
model.save_weights(file_path) #Save model weights to google colab

In [ ]:
model.load_weights(file_path) #Load model weights

In [ ]:
# Fit model with 30 epochs and a batch size of 256
history = model.fit(
    tf_ds,
    batch_size=batch_size,
    epochs = 1,
    callbacks=[checkpoint_callback],
    validation_data = tf_ds_val,

)

2107/2107 ━━━━━━━━━━━━━━━━━━━━ 979s 462ms/step - accuracy: 0.4999 - loss: 0.2780 - top-5-accuracy: 0.9435 - val_accuracy: 0.4704 - val_loss: 0.3033 - val_top-5-accuracy: 0.9133


In [ ]:
file_path = "/content/drive/MyDrive/epoch2.weights.h5" #Create file path another set of weights set of weight

In [ ]:
model.save_weights(file_path) #Save model weights to google colab

In [ ]:
model.load_weights(file_path) #Load model weights

In [ ]:
# Fit model with 30 epochs and a batch size of 256
history = model.fit(
    tf_ds,
    batch_size=batch_size,
    epochs = 1,
    callbacks=[checkpoint_callback],
    validation_data = tf_ds_val,

)

2107/2107 ━━━━━━━━━━━━━━━━━━━━ 980s 463ms/step - accuracy: 0.5001 - loss: 0.2769 - top-5-accuracy: 0.9449 - val_accuracy: 0.4839 - val_loss: 0.3043 - val_top-5-accuracy: 0.9150


In [ ]:
file_path = "/content/drive/MyDrive/epoch3.weights.h5"

In [ ]:
model.save_weights(file_path)

In [ ]:
model.load_weights(file_path)

In [ ]:
# Fit model with 30 epochs and a batch size of 256
history = model.fit(
    tf_ds,
    batch_size=batch_size,
    epochs = 1,
    callbacks=[checkpoint_callback],
    validation_data = tf_ds_val,

)

2107/2107 ━━━━━━━━━━━━━━━━━━━━ 979s 462ms/step - accuracy: 0.5013 - loss: 0.2759 - top-5-accuracy: 0.9465 - val_accuracy: 0.4859 - val_loss: 0.3052 - val_top-5-accuracy: 0.9196


In [ ]:
file_path = "/content/drive/MyDrive/epoch4.weights.h5" #Create file path another set of weights set of weight

In [ ]:
model.save_weights(file_path)

In [ ]:
model.load_weights(file_path)

In [ ]:
tf_ds_test = ds["test"].to_tf_dataset( #Create image generator for Test Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=256,
    num_workers = 48 #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU
)

In [ ]:
# Load best model weights found during Training
model.load_weights(checkpoint_filepath)
_ , accuracy, top_5_accuracy = model.evaluate(tf_ds_test)
print(f"Test accuracy: {round(accuracy * 100, 2)}%")
print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")

492/492 ━━━━━━━━━━━━━━━━━━━━ 196s 299ms/step - accuracy: 0.5152 - loss: 0.3044 - top-5-accuracy: 0.9115
Test accuracy: 51.52%
Test top 5 accuracy: 91.15%


In [ ]:
model.evaluate(tf_ds_test) #Evalute Model

492/492 ━━━━━━━━━━━━━━━━━━━━ 169s 285ms/step - accuracy: 0.5152 - loss: 0.3044 - top-5-accuracy: 0.9115


[0.3043759763240814, 0.5152146220207214, 0.9115487933158875]

In [ ]:
y_predictions = model.predict(tf_ds_test) # Get Predicted # labels from test set
y_true = ds["test"]["labels"]

492/492 ━━━━━━━━━━━━━━━━━━━━ 178s 281ms/step


In [ ]:
import tensorflow as tf
y_predictions_probabiliities = tf.nn.sigmoid(y_predictions) # Get Sigmoid Activation function results from predicted logits

In [ ]:
y_predictions_probabiliities

<tf.Tensor: shape=(125866, 19), dtype=float32, numpy=
array([[0.04464266, 0.04585558, 0.06067286, ..., 0.04310035, 0.35081634,
        0.04678524],
       [0.17222461, 0.05477516, 0.12044106, ..., 0.04848444, 0.59105766,
        0.06455382],
       [0.0517142 , 0.04007398, 0.04655611, ..., 0.04605641, 0.04946726,
        0.9557816 ],
       ...,
       [0.9561364 , 0.10466329, 0.7959422 , ..., 0.04292335, 0.0653334 ,
        0.03789663],
       [0.83387405, 0.06197765, 0.54035974, ..., 0.08152995, 0.04992734,
        0.04855495],
       [0.06611302, 0.03674142, 0.07762849, ..., 0.04498776, 0.07205315,
        0.04862236]], dtype=float32)>

In [ ]:
y_predictions = np.array(y_predictions_probabiliities) # Convert prediction probabilties to numpy array

In [ ]:
y_predictions

array([[0.04464266, 0.04585558, 0.06067286, ..., 0.04310035, 0.35081634,
        0.04678524],
       [0.17222461, 0.05477516, 0.12044106, ..., 0.04848444, 0.59105766,
        0.06455382],
       [0.0517142 , 0.04007398, 0.04655611, ..., 0.04605641, 0.04946726,
        0.9557816 ],
       ...,
       [0.9561364 , 0.10466329, 0.7959422 , ..., 0.04292335, 0.0653334 ,
        0.03789663],
       [0.83387405, 0.06197765, 0.54035974, ..., 0.08152995, 0.04992734,
        0.04855495],
       [0.06611302, 0.03674142, 0.07762849, ..., 0.04498776, 0.07205315,
        0.04862236]], dtype=float32)

In [ ]:
y_predictions_encoded = (y_predictions > 0.5).astype(int) # If probabilites for a label is greater than 0.5, make that label 1

In [ ]:
y_predictions_encoded[3]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0])

In [ ]:
# Calculate Metrics
from sklearn.metrics import precision_score, recall_score, f1_score
# Get micro precision score
print(precision_score(y_true, y_predictions_encoded, average='micro'))
# Get micro recall score
print(recall_score(y_true, y_predictions_encoded, average='micro'))
# Get micro f1 score
print(f1_score(y_true, y_predictions_encoded, average='micro'))


0.25956119256088617
0.23098787840394855
0.2444423717599326


In [ ]:
model.summary() #Model Summary

Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 120, 120,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ data_augmentation   │ (None, 120, 120,  │          0 │ input_layer[0][0] │
│ (Sequential)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ cct_tokenizer       │ (None, 900, 256)  │    149,184 │ data_augmentatio… │
│ (CCTTokenizer)      │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ position_embedding  │ (None, 900, 256)  │    230,400 │ cct_tokenizer[0]… │
│ (PositionEmbedding) │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add (Add)           │ (None, 900, 256)  │          0 │ cct_tokenizer[0]… │
│                     │                   │            │ position_embeddi… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalization │ (None, 900, 256)  │        512 │ add[0][0]         │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 256)  │  1,051,904 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth    │ (None, 900, 256)  │          0 │ multi_head_atten… │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_1 (Add)         │ (None, 900, 256)  │          0 │ stochastic_depth… │
│                     │                   │            │ add[0][0]         │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 256)  │        512 │ add_1[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 900, 256)  │     65,792 │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 900, 256)  │          0 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 900, 256)  │     65,792 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 900, 256)  │          0 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth_1  │ (None, 900, 256)  │          0 │ dropout_2[0][0]   │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_2 (Add)         │ (None, 900, 256)  │          0 │ stochastic_depth… │
│                     │                   │            │ add_1[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 256)  │        512 │ add_2[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 256)  │  1,051,904 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat

 Total params: 22,476,926 (85.74 MB)

 Trainable params: 7,492,308 (28.58 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 14,984,618 (57.16 MB)

# BigEarthNet Classification using CCT

This project demonstrates the classification of satellite images from the BigEarthNet dataset using a Compact Convolutional Transformer (CCT) model implemented with Keras. The BigEarthNet dataset is a large-scale archive of Sentinel-2 satellite images with multi-labels, representing various land cover classes.

## Experiment Overview

The goal of this experiment is to train a CCT model to classify the land cover in the BigEarthNet dataset. The experiment involves:

1.  **Data Loading and Preprocessing**: Loading the BigEarthNet dataset from the Hugging Face Hub, converting the 43-class labels to a 19-class version, and hot-one encoding the labels.
2.  **Model Definition**: Implementing the CCT architecture using custom Keras layers for tokenization, positional embedding, sequence pooling, and stochastic depth.
3.  **Model Training**: Training the CCT model on the preprocessed training data using the AdamW optimizer and Binary Crossentropy loss, with a ModelCheckpoint callback to save the best weights based on validation accuracy.
4.  **Model Evaluation**: Evaluating the trained model on the test set using accuracy, top-5 accuracy, precision, recall, and F1-score.

## Dataset

The BigEarthNet dataset used in this experiment is sourced from the Hugging Face Hub: [danielz01/BigEarthNet-S2-v1.0](https://huggingface.co/datasets/danielz01/BigEarthNet-S2-v1.0). The dataset contains Sentinel-2 satellite images and associated land cover labels. The original 43 classes are mapped to a reduced set of 19 classes for this experiment using the `map_labels` function. The labels are then hot-one encoded using the `hotOne` function.

## Model Architecture

The model used is a Compact Convolutional Transformer (CCT). The key components of the CCT implemented here include:

*   **CCT Tokenizer**: A convolutional network to extract features from image patches and flatten them into sequences.
*   **Position Embedding**: Adds positional information to the tokenized sequences.
*   **Transformer Layers**: Multiple layers of multi-head attention and MLP blocks with stochastic depth for regularization.
*   **Sequence Pooling**: Pools the sequence of tokens into a single representation for classification.

The model hyperparameters are defined as follows:
- Positional Embedding: `True`
- Convolutional Layers: `2`
- Projection Dimension: `256`
- Number of Transformer Heads: `4`
- Number of Transformer Layers: `6`
- Stochastic Depth Rate: `0.1`
- Learning Rate: `0.001`
- Weight Decay: `0.0001`
- Batch Size: `128`
- Number of Epochs: `30`
- Image Size: `120x120`
- Number of Classes: `19`

The model summary shows a total of 22,476,926 parameters, with 7,492,308 trainable parameters.

## Training and Evaluation

The model was trained for 30 epochs using the `tf_ds` and `tf_ds_val` datasets, which are TensorFlow datasets created from the preprocessed BigEarthNet training and validation sets. A `ModelCheckpoint` callback was used to save the best weights based on validation accuracy to `/content/drive/MyDrive/BigEarthFinal/checkpoint.weights.h5`.

After training, the model was evaluated on the test set (`tf_ds_test`) using the best weights saved during training. The evaluation metrics are as follows:

- Test accuracy: 51.52%
- Test top 5 accuracy: 91.15%
- Micro Precision: 0.2596
- Micro Recall: 0.2310
- Micro F1-score: 0.2444

The predicted probabilities for the test set were obtained using `model.predict(tf_ds_test)` and then converted to binary predictions using a threshold of 0.5. These binary predictions were used to calculate the precision, recall, and F1-score.